### PR TITLE
Report mutation test coverage of proposed changes

### PR DESCRIPTION
**Pull request by Stephan202**
Suggested commit message:
```
Report mutation test coverage of proposed changes (#346)

Thanks to a free Arcmutate OSS license, GitHub Actions now runs Pitest against
files changed in the context of a PR. 

While there, update Pitest's configuration to ignore Refaster rule collection
classes, as mutations of those classes will not impact the associated unit
tests.

See:
- https://www.arcmutate.com
- https://pitest.org
```

### CODE REVIEW

File: .github/workflows/pitest-update-pr.yml
Comment (Stephan202) [inline .github/workflows/pitest-update-pr.yml:3]: The changes in this file will only take effect once merged to the default branch. See the code reverted in the most recent commit to see how I tested a variation of this code.

File: pom.xml
Comment (Stephan202) [inline pom.xml:1285]: Arguably this should be moved to a separate PR :shrug:

**Review by Stephan202 — COMMENTED** [review]
Before reviewing this PR, consider reading the corresponding [blog post](https://blog.pitest.org/oss-pitest-pr/).

File: .github/workflows/pitest-analyze-pr.yml
Comment (rickie) [inline .github/workflows/pitest-analyze-pr.yml:5]: To be more similar to the other file we could do `: analysis` here.

File: .github/workflows/pitest-analyze-pr.yml
Comment (rickie) [inline .github/workflows/pitest-analyze-pr.yml:27]: Seems fun to me to add some values for `killedEmoji` and `mutantEmoji`. Will propose something 👀. 

Relevant documentation [here](https://docs.arcmutate.com/docs/github/github-maven-plugin.html).

**Review by rickie — COMMENTED** [review]
Started reviewing this one, cool stuff!

Added a commit.

This might break things, let's see later when testing the changes 👀.

**Comment by rickie** [comment]
Rebased :).

File: cdg-pitest-license.txt
Comment (Stephan202) [inline cdg-pitest-license.txt:7]: @rickie following your last commit the codes doesn't work anymore; the plugin looks for the exact original file. Now all analysis is skipped. From the logs:
```
Licenced for evaluation and demonstration purposes only.
Expires: 2023-09-20
For use with packages: com.example.*
```

File: .github/workflows/pitest-analyze-pr.yml
Comment (Stephan202) [inline .github/workflows/pitest-analyze-pr.yml:5]: I had that `: analysis` suffix in an earlier commit, and I removed it on purpose, IIRC. Something to do with how the result is displayed in the GitHub UI. (But would have to revert the last commit to see how/what.)
```suggestion
name: "Mutation testing"
```

**Review by Stephan202 — COMMENTED** [review]
I'll let you do the revert and test the other changes :upside_down_face:

File: cdg-pitest-license.txt
Comment (Stephan202) [inline cdg-pitest-license.txt:1]: To be sure: I'd also revert the change to this file. It doesn't follow our standards anyway (missing dot at end of sentence, auto-generated timestamp), so I'd rather keep it exactly as provided.

Comment (Stephan202) [reply cdg-pitest-license.txt:1]: Or we should replace all comments with something like:
```
# Arcmutate license for Error Prone Support, requested by sending an email to
# support@arcmutate.com.
```
(Assuming the plugin doesn't care about the comments :eyes:)

File: pom.xml
Comment (rickie) [reply pom.xml:1248]: It's okay to leave in IMO.

File: .github/workflows/pitest-update-pr.yml
Comment (rickie) [inline .github/workflows/pitest-update-pr.yml]: Here we should use the same version as in the existing workflow `17.0.4`, right?

**Review by rickie — APPROVED** [review]
Added a commit. Also reverted the test changes. 

The setup looks good to me, very cool to have this as part of the build 😄.

Doubting a little about the suggested commit message, we could provide some more information about Arcmutate and the setup we now have. I'm thinking there is a reason that you kept it high-level, so first asking whether there is a particular reason for it 👀 @Stephan202.

**Comment by Stephan202** [comment]
Latest changes LGTM! Also updated the suggested commit message. PTAL :)

**Comment by rickie** [comment]
Nice, looks splendid 😄!

File: .github/workflows/pitest-analyze-pr.yml
Comment (Badbond) [inline .github/workflows/pitest-analyze-pr.yml:25]: At first, the `+GIT(from[HEAD~1])` triggered me thinking this would only apply on the latest commit of the feature branch, and that we instead wish to diff between the head of the target and the head of the feature branch. However, this actually refers to the entire diff as it is squashed into a single commit (see [here](https://blog.pitest.org/pitest-pr-setup/#:~:text=This%20runs%20pitest%2C%20producing%20additional%20output%20which%20is%20then%20used%20by%20the%20maven%20plugin.%20Although%20at%20first%20glance%20it%20looks%20like%20only%20the%20most%20recent%20change%20is%20analysed%2C%20in%20fact%20all%20changes%20in%20the%20PR%20will%20be%20considered%20as%20github%20places%20them%20in%20a%20single%20commit.)):

> This runs pitest, producing additional output which is then used by the maven plugin. Although at first glance it looks like only the most recent change is analysed, in fact all changes in the PR will be considered as github places them in a single commit.

Perhaps worth of a comment to avoid future confusion.

**Review by Badbond — APPROVED** [review]
Looking good. Nice suggested commit message too. Didn't dive to deep into all the possbile configuration to see if we can improve further. I've got two comments, but will approve nonetheless as they are not blocking this PR for me.

File: .github/workflows/pitest-analyze-pr.yml
Comment (rickie) [reply .github/workflows/pitest-analyze-pr.yml:13]: Fair point :). 

However, somewhat regularly we have Renovate updates, like for Pitest or its JUnit plugin, where we would want to see that the setup still works with the new version. So I think it's better to leave it as is 😄.

File: .github/workflows/pitest-analyze-pr.yml
Comment (rickie) [reply .github/workflows/pitest-analyze-pr.yml:25]: The blogpost is mentioned in a comment on the second line, so I'm in doubt whether we need a comment on top of that. Both options are fine by me though.

Comment (Badbond) [reply .github/workflows/pitest-analyze-pr.yml:25]: What I am linking here is a different blog post though.

Comment (Badbond) [reply .github/workflows/pitest-analyze-pr.yml:28]: Added a commit. PTAL.

**Comment by Stephan202** [comment]
^ Will review the latest discussions EOD.

File: .github/workflows/pitest-analyze-pr.yml
Comment (rickie) [reply .github/workflows/pitest-analyze-pr.yml:28]: LGTM! Nice.

Comment (Badbond) [reply .github/workflows/pitest-analyze-pr.yml:30]: Very nice. Much more details but still concise. :+1: LGTM.

**Pull request merged**
